src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -21,133 +21,505 @@
+    public static final long MIN_PROBE_INTERVAL_MS = Long.getLong("cassandra.dynamic_snitch_min_probe_interval_ms", 60 * 1000L) ;
+    // The probe rate is set later when configuration is read in applyConfigChanges
+    protected static final RateLimiter probeRateLimiter = RateLimiter.create(1);
+    protected static final DebuggableScheduledThreadPoolExecutor latencyProbeExecutor = new DebuggableScheduledThreadPoolExecutor("LatencyProbes");
+    private long lastUpdateSamplesNanos = System.nanoTime();
+
+    // User configuration of the snitch tunables
+    protected volatile int dynamicUpdateInterval = -1;
+    protected volatile int dynamicSampleUpdateInterval = -1;
+    protected volatile double dynamicBadnessThreshold = 0;
 
     // the score for a merged set of endpoints must be this much worse than the score for separate endpoints to
     // warrant not merging two ranges into a single range
     private static final double RANGE_MERGING_PREFERENCE = 1.5;
 
     private String mbeanName;
-    private boolean registered = false;
-
-    private volatile HashMap<InetAddressAndPort, Double> scores = new HashMap<>();
-    private final ConcurrentHashMap<InetAddressAndPort, ExponentiallyDecayingReservoir> samples = new ConcurrentHashMap<>();
+    private boolean mbeanRegistered = false;
 
     public final IEndpointSnitch subsnitch;
 
-    private volatile ScheduledFuture<?> updateSchedular;
-    private volatile ScheduledFuture<?> resetSchedular;
-
-    private final Runnable update;
-    private final Runnable reset;
+    private volatile ScheduledFuture<?> updateScoresScheduler = null;
+    private volatile ScheduledFuture<?> updateSamplesScheduler = null;
 
     public DynamicEndpointSnitch(IEndpointSnitch snitch)
     {
         this(snitch, null);
     }
 
-    public DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
+    protected DynamicEndpointSnitch(IEndpointSnitch snitch, String instance)
     {
         mbeanName = "org.apache.cassandra.db:type=DynamicEndpointSnitch";
         if (instance != null)
             mbeanName += ",instance=" + instance;
         subsnitch = snitch;
-        update = new Runnable()
+
+        if (DatabaseDescriptor.isDaemonInitialized())
         {
-            public void run()
+            open();
+        }
+    }
+
+    /**
+     * Update configuration of the background tasks and restart the various scheduler tasks
+     * if the configured rates for these tasks have changed.
+     */
+    public void applyConfigChanges(int newDynamicUpdateInternal, int newDynamicSampleUpdateInterval, double newDynamicBadnessThreshold)
+    {
+        if (DatabaseDescriptor.isDaemonInitialized())
+        {
+            if (dynamicUpdateInterval != newDynamicUpdateInternal || updateScoresScheduler == null)
             {
-                updateScores();
+                cancelAndWait(updateScoresScheduler, Math.max(1, dynamicUpdateInterval), TimeUnit.MILLISECONDS);
```

Comment:
Interesting, I usually opt to either not block or block with a timeout, but I see your point. In this case none of the futures that are being called should block, and it can't block during the initial startup (at constructor time the futures are `null` so they just immediately succeed at cancelling) so I think I'll just block forever although it pains me to do so heh.

I tried just weaving the TimeoutException back up the stack so we could have our cake and eat it too, but there are lots of places that handling it just becomes difficult (e.g. in the backwards compatible mbean interface I'd have to re-throw it as a runtime exception or something which is confusing and not clear that your DES is now in a bad state).